src/multisig/multisig_tx_builder_ringct.cpp

```diff
@@ -0,0 +1,936 @@
+      return false;
+  }
+
+  // prepare rings for input CLSAGs
+  rv.mixRing.resize(num_sources);
+  for (std::size_t i = 0; i < num_sources; ++i) {
+    const std::size_t ring_size = sources[i].outputs.size();
+    rv.mixRing[i].resize(ring_size);
+    for (std::size_t j = 0; j < ring_size; ++j) {
+      rv.mixRing[i][j].dest = sources[i].outputs[j].second.dest;
+      rv.mixRing[i][j].mask = sources[i].outputs[j].second.mask;
+    }
+  }
+
+  // make pseudo-output commitments
+  rct::keyV a;  //pseudo-output commitment blinding factors
+  auto a_wiper = epee::misc_utils::create_scope_leave_handler([&]{
+    memwipe(static_cast<rct::key *>(a.data()), a.size() * sizeof(rct::key));
+  });
+  if (not reconstruction) {
+    a.resize(num_sources);
+    rv.p.pseudoOuts.resize(num_sources);
+    a[num_sources - 1] = rct::zero();
+    for (std::size_t i = 0; i < num_destinations; ++i) {
+      sc_add(
+        a[num_sources - 1].bytes,
+        a[num_sources - 1].bytes,
+        output_amount_masks[i].bytes
+      );
+    }
+    for (std::size_t i = 0; i < num_sources - 1; ++i) {
+      rct::skGen(a[i]);
+      sc_sub(
+        a[num_sources - 1].bytes,
+        a[num_sources - 1].bytes,
+        a[i].bytes
+      );
+      rct::genC(rv.p.pseudoOuts[i], a[i], sources[i].amount);
+    }
+    rct::genC(
+      rv.p.pseudoOuts[num_sources - 1],
+      a[num_sources - 1],
+      sources[num_sources - 1].amount
+    );
+  }
+  // check balance if reconstructing the tx
+  else {
+    rv.p.pseudoOuts = unsigned_tx.rct_signatures.p.pseudoOuts;
+    if (num_sources != rv.p.pseudoOuts.size())
+      return false;
+    rct::key balance_accumulator = rct::scalarmultH(rct::d2h(fee));
+    for (const auto& e: rv.outPk)
+      rct::addKeys(balance_accumulator, balance_accumulator, e.mask);
+    for (const auto& pseudoOut: rv.p.pseudoOuts)
+      rct::subKeys(balance_accumulator, balance_accumulator, pseudoOut);
+    if (not (balance_accumulator == rct::identity()))
+      return false;
+  }
+
+  // prepare input CLSAGs for signing
+  const rct::key message = get_pre_mlsag_hash(rv, hw::get_device("default"));
+
+  rv.p.CLSAGs.resize(num_sources);
+  if (reconstruction) {
+    if (num_sources != unsigned_tx.rct_signatures.p.CLSAGs.size())
+      return false;
+  }
+
+  CLSAG_contexts.resize(num_sources);
+  if (not reconstruction)
+    cached_w.resize(num_sources);
+
+  for (std::size_t i = 0; i < num_sources; ++i) {
+    const std::size_t ring_size = rv.mixRing[i].size();
+    const rct::key& I = sources[i].multisig_kLRki.ki;
+    const std::size_t l = sources[i].real_output;
+    if (l >= ring_size)
+      return false;
+    rct::keyV& s = rv.p.CLSAGs[i].s;
+    const rct::key& C_offset = rv.p.pseudoOuts[i];
+    rct::keyV P(ring_size);
+    rct::keyV C_nonzero(ring_size);
+
+    if (not reconstruction) {
+      s.resize(ring_size);
+      for (std::size_t j = 0; j < ring_size; ++j) {
+        if (j != l)
+          s[j] = rct::skGen();  //make fake responses
```

Comment:
This can be converted into a RNG in a manner similar to tx_secret_key to reduce bandwidth and consolidate code between creator/reconstructors.

Reply:
I agree you _can_ do this, and it does reduce bandwidth. I'm not sure it is useful in practice, because the payload size would only shrink by a few hundred bytes at most (in the typical case). The goal here is to get something that works correctly, rather than engineer it into the ideal state.